### DISCUSSION TITLE
Why does vec4<u32> >>= not work but vec4<u32> /= u32 does?

### DISCUSSION THREAD

**Discussion by jacobgorm** [gpuweb/gpuweb#4750] [Q&A]
This does not work:
```
var myvector: vec4<u32>
var shift = 1u;
myvector >>= shift;
```
But this does

```
var myvector: vec4<u32>
var shift = 1u;
var divs = 1u << shift;
myvector /= divs;
```

Testing with latest tint compiler built from source.

Can anybody explain why shifting a vector of u32's with a u32 would not work, while division does? What are the odds of the division getting optimized back into a shift?

**Comment by mehmetoguzderin** [comment]
As far as I can see in the compound assignment section, both `>>=` and `/=` seem to have a common treatment: https://gpuweb.github.io/gpuweb/wgsl/#compound-assignment-sec

But in expressions which assignments decompose into, the arithmetic section contains mixed vector-scalar operations while bitwise does not contain: https://gpuweb.github.io/gpuweb/wgsl/#arithmetic-expr

There was an instance of including modulo later as an operation that can take in mixed vector and scalar: https://github.com/gpuweb/gpuweb/pull/2495

Also relevant: https://github.com/gpuweb/gpuweb/issues/4126

So, although I'm not very familiar with inner workings of Tint, it might be reflecting the spec accurately as of today in terms of this restriction. Though taking scalars on right hand side of bit shift operations sound reasonable to a certain extent.

**Comment by jacobgorm** [comment]
This seems like a bug to me, if one has to go via division to obtain the effect of a switch. Should I file an issue with this?

**Reply by dj2** [reply]
You don't need to go through a division, you just have to shift by a vector.

```
var myvector: vec4<u32>
var shift = vec4(1u);
myvector >>= shift;
```

**Reply by mehmetoguzderin** [reply]
@jacobgorm I think it makes sense to file an issue if you feel it is inconvenient in comparison to other shading languages to need an explicit conversion of a scalar to vector before using shift assignment operator so that it is on the record.

**Reply by greggman** [reply]
Where in grammar is this covered?

This works

```
  var a: vec4i = vec4i(0);
  a += 1;
```

This doesn't

```
  var a: vec4i = vec4i(0);
  a >>= 1;
```

At a glance, I don't see where [_shift_right_assign](https://gpuweb.github.io/gpuweb/wgsl/#syntax_sym-_shift_right_assign) and
 [_shift_left_assign](https://gpuweb.github.io/gpuweb/wgsl/#syntax_sym-_shift_left_assign) which are [compound_assignment_operators](https://gpuweb.github.io/gpuweb/wgsl/#syntax-compound_assignment_operator) are called out as behaving different from the other [compound_assignment_operators](https://gpuweb.github.io/gpuweb/wgsl/#syntax-compound_assignment_operator)

**Reply by mehmetoguzderin** [reply]
@greggman That's right. I pointed out here, it is the mixed vector scalar ops table in arithmetic section: https://github.com/gpuweb/gpuweb/discussions/4750#discussioncomment-10053643

**Reply by dj2** [reply]
The compound ops are specifically called out as being sugar, so `a >>= b` is actually `a = a >> b`. You then have to go lookup the `>>` operator in https://gpuweb.github.io/gpuweb/wgsl/#bit-expr where it lists the operands to `>>` as:

```
e1 >> e2: T

e1: T
e2: TS
S is i32 or u32
T is S or vecN<S>
TS is u32 when T is S, otherwise TS is vecN<u32>
```

So, basically they're spec'd as having to match in dimension.

The other operators like `*` and `/` have a special call out in the `Binary arithmetic expressions with mixed scalar and vector operands` table which provides the (scalar, vector) operations.

Since `>>` and `<<` are bit shift operators, not binary operators, they didn't get the special treatement.

**Comment by jacobgorm** [comment]
This is very inconvenient when trying to create a code base that cross-compiles a battery of existing shaders, and overall hampers code readability because you need to create an extra variable to hold all the shift values. Doing this the "normal way" seems to work everywhere else (C with GCC vector extensions, C++/CUDA with operator overloads, OpenCL, Metal, HLSL). Just hit the same problem with clamp(a, b, c) where a is a vector and b and c are scalars.

**Reply by mehmetoguzderin** [reply]
@jacobgorm you can file issues for the conversion inconveniences you come across https://github.com/gpuweb/gpuweb/discussions/4750#discussioncomment-10089382

**Reply by mehmetoguzderin** [reply]
Or I can try to help in case you'd prefer me to file them.

**Reply by jacobgorm** [reply]
You are very welcome if you want to and know the right wording :-)

**Reply by mehmetoguzderin** [reply]
@jacobgorm sure thing! I tried to file them, please feel free to add as comment if you feel like I could not represent the concerns well at some points.

1. Mixed bitshift assign: https://github.com/gpuweb/gpuweb/issues/4816
2. Mixed clamp: https://github.com/gpuweb/gpuweb/issues/4815

**Comment by jacobgorm** [comment]
Of course (as I just found out) the same problem exists for min() and max() and probably a lot of other functions that one would use for, e.g., activations in a neural network.
